### ISSUE TITLE
Mocking higher order functions

### ISSUE THREAD

**Issue by THEStephenStanton** [paulbutcher/ScalaMock#231]
### Scala Version 2.11.7

I need some help with how to mock higher order functions inside classes. 

As you can see with my code below, the function being mocked works fine when you pass in a value that has the higher order function, but won't if you type it out each spot.

This is a simplified example, but I need a way to get the former to work.  Reason being (I'll try my best to explain this) I have a class A that is being tested.  Inside A is a function that is passed a mocked class B, and basically the foo function (as seen below) is inside this mocked B and I can't just pass in f like I do in the second example below.  If this didn't make any sense, I can try to replicate this exactly.

TL;DR I need the first test to work lol

Any ideas why this might be happening? 

Please provide a minimised code snippet that fails, ideally, written as a failing test case in ScalaTest.
This will help us a lot in diagnosing the problem and work on a fix.
If the issue is more complex or requires configuration, please provide a link to a project on Github that reproduces the issue.

```scala
  class Foo {
    def foo(f: () ⇒ String) = "this is foo"
  }

  val fooMock = mock[Foo]

  "Special test" should "pass but doesn't" in {
    (fooMock
      .foo(_: () ⇒ String))
      .expects({ () ⇒
        "bar"
      })
      .returns("this is the test")

    // will fail here
    val result = fooMock.foo({ () ⇒
      "bar"
    })

    assert(true)
  }

  "Another Test" should "that passes but I can't do it this way" in {
    val f = { () ⇒
      "bar"
    }

    (fooMock.foo(_: () ⇒ String)).expects(f).returns("this is the test")

    val result = fooMock.foo(f)

    result shouldBe "this is the test"
  }
```

**Comment by THEStephenStanton** [comment]
To be more clear, here is a more accurate test of what I am trying to accomplish

```scala
  class Foo {
    def foo(f: () ⇒ String) = s"you sent in: ${f()}"
  }

  object Bar {

    def bar(foo: Foo): String = {
      val f = { () ⇒ "bar" }

      foo.foo(f)
    }
  }

  val fooMock = mock[Foo]

  "Special test" should "pass but doesn't" in {
    (fooMock.foo(_: () ⇒ String))
      .expects({ () ⇒
        "bar"
      })
      .returns("this is the test")

    // will fail here
    val result = Bar.bar(fooMock)

    assert(true)
  }

```

**Comment by barkhorn** [comment]
see http://scalamock.org/user-guide/matching/ for predicate matching, which is I think what you are trying to do.

It doesn't look like you are trying to report a bug though, so I'll close this. Requests for help are better placed on Stackoverflow, not here.
